enhancements/local-storage/automatic-detection.md

```diff
@@ -64,40 +65,299 @@ and we can use that information to sort devices into storage classes based on th
+	DiscoveredTimeStamp string                         `json:"discoveredTimeStamp,omitempty"`
+	// Devices contains the list of devices on which LSO
+	// is capable of creating LocalPVs
+	// The devices in this list qualify these following conditions.
+	// - it should be a non-removable device.
+	// - it should not be a read-only device.
+	// - it should not be mounted anywhere
+	// - it should not be a boot device
+	// - it should not have child partition in lsblk output
+	DiscoveredDevices []DiscoveredDevice `json:"discoveredDevices"`
+}
+
+type LocalVolumeDiscoveryResult struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              LocalVolumeDiscoveryResultSpec   `json:"spec"`
+	Status            LocalVolumeDiscoveryResultStatus `json:"status,omitempty"`
+}
+
+type LocalVolumeDiscoveryResultList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []LocalVolumeDiscoveryResult `json:"items"`
+}
+
+```
+
+### Workflow of `Auto Device Discovery`
+The discovery of local devices will be done by the `LocalVolumeDiscovery` and `LocalVolumeDiscoveryResult` CR.
+
+#### LocalVolumeDiscovery Controller
+- Once the `LocalVolumeDiscovery` CR is deployed, its controller will:
+  - Create the `LocalVolumeDiscoveryResult` CRD if its not available.
+  - Create a daemonSet to run on each of the valid nodes.(Information about the valid list of nodes would be available in the `LocalVolumeDiscovery` CR as `nodeSelector`)
+  - The `Daemonset` should have `LocalVolumeDiscovery` CR as owner reference.
+  - Controller would remember the current timestamp when the discovery was triggered. (say `discoveryStarted`)
+  - Update the `Phase` in the `LocalVolumeDiscovery` CR to `Discovering`.
+- The DaemonSet pod running on each valid node will:
+  - Create `LocalVolumeDiscoveryResult` CR on that node if not present.
+  - Identify the list of devices available on that node.
+  - Update the `status` of the `LocalVolumeDiscoveryResult` CR with the list of devices identified above.
+  - Update the `status` of the `LocalVolumeDiscoveryResult` with `discoveredTimeStamp`. This represents the timestamp when the discovery was completed.
+`
+#### `Auto Device Discovery` Status
+- The `LocalVolumeDiscovery` controller will track the status of the disovery process by:
+  - Tracking the `LocalVolumeDiscoveryResult` CR for each node. For `n` nodes, the controller should ensure that `n` `LocalVolumeDiscoveryResult` CR should be created.
+  - Once all the `LocalVolumeDiscoveryResult` CR are created, the controller would check `discoveredTimeStamp` timestamp for each `LocalVolumeDiscoveryResult` CR.
+  - When the `discoveredTimeStamp` timestamp for each `LocalVolumeDisocveryResult`CR is greater than `discoveryStarted` timestamp, then update the `Phase` in the `LocalVolumeDiscovery` CR to `Discovered`.
+- Controller should update the `LocalVolumeDiscovery` `Phase` as `Error` if:
```

Comment:
I still don't see the value in updating the discovery status globally like this when it's for ongoing discovery. It might or might not be considered an error that a node didn't respond. Maybe it's expected that a node is offline. The higher level entity would need to determine if it's an error for discovery not to be completed or if the discovery timestamp is too old. For example, OCS would decide if it can continue to the next configuration step even if all nodes are not responding.

Reply:
One idea to use the `phase` like this is to let the applications using LSO (the OCP UI) to know when the discovery process is completed (so that it can move on to the next step, that is, `auto-provisioning`).  But yes, it won't make much sense with continuous discovery.
1. Should we rather have the users (OCP UI) to reply on the individual `LocalVolumeDiscoveryResult` CR to track the completion of discovery?
2. Or any other efficient way to track the progress of the discovery for the users ?

Reply:
> Should we rather have the users (OCP UI) to reply on the individual LocalVolumeDiscoveryResult CR to track the completion of discovery?

Yes, this is what seems necessary. The consumer of the LocalVolumeDiscoveryResult CRs will need to decide if they have enough information, that the discovery is recent enough, that it can skip nodes that aren't responding, etc.